src/style.rs

```diff
@@ -674,12 +678,96 @@ enum Alignment {
     Right,
 }
 
+/// Trait for defining stateful or stateless formatters
+pub trait ProgressTracker: Send {
+    fn spawn(&self) -> Box<dyn ProgressTracker>;
+    fn tick(&mut self, state: &ProgressState, now: Instant);
+    fn reset(&mut self, state: &ProgressState, now: Instant);
+    fn write(&self, state: &ProgressState, w: &mut dyn fmt::Write) -> fmt::Result;
+}
+
+impl Clone for Box<dyn ProgressTracker>
+where
+    Self:,
+{
+    fn clone(&self) -> Self {
+        self.spawn()
+    }
+}
+
+impl<F: Fn(&ProgressState) -> String + Send + Clone + 'static> ProgressTracker for F {
```

Comment:
Can we change the `Fn` here to take the `fmt::Write`, too?

Reply:
I mean, currently we have `F` return a `String`. We changed the trait method to write into a `fmt::Formatter<'_>` instead, can we change the `F` bound to be similar (that is, avoid allocating a `String` here)?